[4-6단계 - 방탈출 사용자 예약] 밍곰(박민선) 미션 제출합니다.

minSsan opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  [방탈출 사용자 예약](https://prolog.techcourse.co.kr/studylogs/4288)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
이번 미션에서는 완성에 더 집중해서 그만큼 생활체조 규칙을 신경쓰지 못했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

## 이전 PR 코멘트 답변

https://github.com/woowacourse/spring-roomescape-member/pull/208#discussion_r2075382524
> ... 객체의 상태가 꽤 많아질 수록 검증할 요소나 실수로 잘못 주입한다거나 등의 상황이 발생 혹은 생성할 수 있는 케이스가 다양할 수도 있다고 생각하는데요. 이런 모든 경우의 수를 생성자만으로는 해결할 수 없다고 생각해요.
> 다만, 객체지향에서의 객체의 개념과 관계형 데이터 베이스의 패러다임이 불일치하다보니 이런 불일치적인 요소 때문에 도메인 객체 생성 여부 판단을 객체가 스스로 판단하지 못하는 경우가 생기고 이런 경우에는 서비스단에서 검증을 할 수 밖에 없다고 생각해요.

어플리케이션 코드에서는 **과거에 생성된 예약 내역도 `Reservation` 객체로 다뤄야함**에 따라, `Reservation` 객체의 생성자에서 예약 생성 조건 _(= 과거 날짜 불가능)_ 을 검증하는 것이 불필요한 것 같다는 생각이 드네요..! 

**예약 생성 조건**은 **사용자가 예약을 새로 생성(create)할 때만 필요**한 조건이기 때문에, `Reservation` 생성자보다는 create 요청시에 호출되는 **서비스 레이어에서 검증하는 것**이 좋을 것 같아요~!

기존에 서비스 레이어에서 검증했던 방식으로 다시 변경했습니다! 

[feat: 예약 생성 검증 - Reservation -> service 계층으로 변경](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/5b55dc9d1c566bd1cffcf3827bc5c747f351f262)
[feat: 시간 생성 검증 - ReservationTime -> service 계층으로 변경](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/837fff52521ec175b412dc6b742383d930a3e7ab)

minSsan committed:
feat: 로그인 뷰 렌더링

minSsan committed:
🔴red: 로그인 요청시 set-cookie로 토큰을 응답할 수 있다 테스트

minSsan committed:
feat(build.gradle): jjwt 의존성 추가

minSsan committed:
🟢green: login시 set-cookie로 토큰 발급 통과 로직 작성

minSsan committed:
refactor(JwtTokenProvider): JWT 토큰 만료기한 설정

minSsan committed:
🔴red: 존재하지 않는 유저 로그인 요청 예외 테스트

minSsan committed:
feat(JwtTokenProvider): TokenProvider 인터페이스 implements 추가

minSsan committed:
🟢green: 존재하지 않는 유저 예외 통과 코드

minSsan committed:
feat: UnAuthorization Exception 핸들러 추가

minSsan committed:
fix: FakeTokenProvider 컴파일 에러 - 네이밍 변경

minSsan committed:
feat: FakeUserRepository 메서드 구현

minSsan committed:
feat(AuthViewController): 회원가입 뷰 연결

minSsan committed:
🔴red: 회원가입 & 로그인

minSsan committed:
🟢green: 회원가입 & 로그인 테스트 통과 로직 작성

minSsan committed:
fix(AuthController): jakarta 타입 검증 @Valid 어노테이션 추가

minSsan committed:
test(UserAuthService): 회원가입 - 중복 이메일 존재 케이스 추가

minSsan committed:
fix(JdbcUserRepository): Select 쿼리 필드 미지정

minSsan committed:
fix(JwtTokenProvider): JWT 디코딩 오류 - JWS로 디코딩함

minSsan committed:
feat: GET /login/check API 구현

minSsan committed:
refactor: 유저 관련 커스텀 예외 추가

minSsan committed:
refactor: auth dto - request / response 분리

minSsan committed:
refactor: 헤더 쿠키에서 토큰을 추출하여 핸들러 메서드로 넘겨주는 HandlerMethodArgumentResolver 추가

minSsan committed:
test: 로그인 테스트 케이스 추가 및 테스트 생명주기 변경 (메서드 -> 클래스)

minSsan committed:
chore: TODO 주석 추가

minSsan committed:
refactor: AuthController 분리 -> Login, Signup

minSsan committed:
refactor: User entity -> Member entity 이름 변경

minSsan committed:
feat: TokenProvider 인터페이스 삭제 및 의존성 테스트를 stub으로 변경

- Mockito 라이브러리 의존성 추가

minSsan committed:
refactor: 불필요한 주석 삭제 및 JwtTokenProvider service -> infrastructure 이동

minSsan committed:
feat: html 4단계 주석 삭제

minSsan committed:
chore: TODO 주석 추가 - repository, service 에러 핸들링 관련

minSsan committed:
feat: Member에 Role 추가

minSsan committed:
feat: reservation 테이블 - name 필드 삭제 및 member_id 필드 추가

minSsan committed:
feat: reservation 생성시 토큰을 기준으로 예약자 판단

minSsan committed:
chore: CheckResponse -> LoginCheckResponse

minSsan committed:
feat: 어드민 전용 Reservation 생성 API 추가 (POST /admin/reservations)

minSsan committed:
feat(schema.sql): time table start_at 필드 타입 변경 - VARCHAR -> TIME

minSsan committed:
test: 테스트 환경 추가

minSsan committed:
feat: reservation response - 예약자명으로 변경

minSsan committed:
refactor: Reservation, ReservationTime 생성자 검증 메서드 삭제

minSsan committed:
refactor: response 정적 생성자 메서드명 수정

minSsan committed:
feat: 어드민 페이지 접근 권한 확인 - Interceptor

minSsan committed:
refactor: Configuration에서 인스턴스를 새로 생성하지 않고 빈을 주입하도록 변경

minSsan committed:
feat: 예약 목록 필터링 추가 (GET /reservations with RequestParam)

minSsan committed:
chore: 불필요한 import 삭제

minSsan committed:
refactor: 패키지 구조 변경

minSsan committed:
test: 테스트 요구사항 변경에 맞게 수정

minSsan committed:
test(FakeReservationRepository): 필터링 기능 구현

minSsan committed:
feat: ErrorResponse 객체 생성 - 예외 응답 포맷

minSsan committed:
feat: 예약 생성 검증 - Reservation -> service 계층으로 변경

minSsan committed:
feat: 시간 생성 검증 - ReservationTime -> service 계층으로 변경

minSsan committed:
chore: 불필요한 변수 선언 삭제

minSsan committed:
chore: 불필요한 주석 삭제

minSsan committed:
feat: LoginMember(entity) -> LoginMemberResponse(dto) 변경

minSsan committed:
feat: LoginMemberResponse -> LoginMember 네이밍 변경

minSsan committed:
feat: MemberTokenMethodArgumentResolver, AuthenticationPrincipal 삭제

Rok93 commented on build.gradle:
```diff
@@ -21,12 +21,16 @@ dependencies {
 
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     testImplementation 'io.rest-assured:rest-assured:5.3.1'
+    testImplementation "org.mockito:mockito-core:5.+"
```
mockito 라이브러리 의존성을 추가하셨군요? 
mockito 라이브러리 의존성이 필요하신 이유가 있을까요?  😃

정말 어쩔 수 없이 써야하는 경우라면 사용해도 좋지만 가능하다면 mockito 라입러리에 의존하지 않고 작성해보면 어떨까요?

minSsan replied:
저도 고민했던 부분인데요 ,,! `JwtTokenProvider`에 의존하는 `MemberService` 객체를 테스트하는 과정에서 `mockito` 의존성을 추가하게 되었어요.

처음에는 `Mock`이 아닌 `Fake 객체`로 처리하려고 했는데, 실제 프로덕션 코드에서 추상화가 필요하지 않음에도 `Fake 객체` 생성을 위해 인터페이스를 정의하는게 맞는건가? 싶은 생각이 들었어요.

그래서 아래와 같이 `JwtTokenProvider`를 `mock`으로 정의했었는데요 _(다시 생각해보니 mock 보다는 stub이 적합할 것 같긴 합니다 ...)_

https://github.com/woowacourse/spring-roomescape-member/blob/786c6cd38f267f0548e6ae71945f934d9e43652f/src/test/java/roomescape/auth/service/MemberServiceTest.java#L18-L22

객체를 인터페이스로 추상하지 않은 경우에는 `mock` 또는 `stub`을 어떻게 적용할 수 있을지, 고민을 많이 안 하고 `mockito` 의존성을 바로 추가했던 것 같아요. 
생각해보니 이 경우에는 **상속**으로도 충분히 `stub` 객체를 구현할 수 있을 것 같아서 내용을 적용해봤어요..! _(이게 맞는 방식인지 모르겠습니다 🥹)_

[feat: mockito 의존성 삭제 및 JwtTokenProvider Stub 상속으로 구현](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/4a3da0dd496507061e1b7adeb6d80e714207e7e5)

### Mock 테스트 기준
실제 테스트 데이터를 기반으로 전반적인 동작을 확인하고 싶을 때 `Fake 객체`를 활용하고, 
> 예를들어 레파지토리는 테스트 데이터를 기반으로 CRUD 작업을 포함하고 있는데요, 이러한 레파지토리는 테스트 코드 내부에서 실제 DB가 동작하는 것처럼 _(= 삭제 요쳥을 하면 삭제가 되고, 생성 요청을 하면 생성이 되는 ...)_ 가짜 객체를 만드는게 좋다고 생각했어요.

주요 도메인 로직을 포함하지 않고, 단순히 값을 생성하는 `JwtTokenProvider` 같은 객체는 Fake 보단 **stub**으로 리턴 값만 고정시켜도 괜찮다고 생각하는데요,

조금 추상적인 기준인 것 같기도 해서 ... 혹시 테스트에서 로키만의 Mock 사용 기준이 따로 있을까요?!

Rok93 replied:
> 생각해보니 이 경우에는 상속으로도 충분히 stub 객체를 구현할 수 있을 것 같아서 내용을 적용해봤어요..! (이게 맞는 방식인지 모르겠습니다 🥹)

상속 방식은 사실 크게 고민해본적이없는데, 말씀하신 것처럼 굳이 추상화할 필요가 없는 케이스를 굳이 인터페이스로 뽑을 필요가 없을 것 같다고 생각해요. 그런 경우에는 별도 인터페이스 없이 테스트만을위한 상속 방식도 괜찮은 방식이지 않을까?  생각되네요. 😃

테스트에서 **Mock/Fake/Stub을 선택하는 기준은 “이 객체의 어떤 행동이 중요하냐?”**에 따라 결정하면 될 것 같아요. 

사실 mockito가 제공해주는 mock이 말로만 mock이지 사실 stub의 역할도할 수 있다보니 가끔은 어떤 Test Double을 쓸지 크게 고민하지 않고 쓰는 경우도 많은 것 같아요. 😅 

저만의 Test Double을 사용 방식을 고민해봤을 때, 아래와 같이 정의해볼 수 있을 것 같아요 🤔 

* Stub: 반환값만 필요하고, 검증할 동작이 없는 경우
* Fake: 상태/로직이 있고 실제 동작과 유사하게 동작해야하는 경우
* Mock: 협력 객체의 메서드 호출 여부나 순서를 검증할 때

Rok93 commented on src/main/java/roomescape/global/infrastructure/JwtTokenProvider.java:
```diff
@@ -0,0 +1,38 @@
+@Component
+public class JwtTokenProvider {
+    @Value("${security.jwt.token.secret-key}")
+    private String SECRET_KEY;
```
SECRET_KEY와 EXPIRE_LENGTH_MILLI는 상수 네이밍을 사용하셨네요. 👀
이 두 값은 JwtTokenProvider 객체의 상태 값으로 봐야하지 않을까요? 😃

minSsan replied:
`@Value` 어노테이션으로 **설정 파일에 있는 값을 주입**해주고 있기도 하고, 객체가 스프링 빈으로 관리되고 있어서 (= **싱글톤 객체**) `secret_key`와 `expire_milli` 값을 상수로 볼 수 있지 않을까 생각했었어요!

인스턴스가 갖는 고유한 **상태 값** 보다는, `빈 등록 시점에서 값이 결정되고 이후에 변하지 않는다는 점`에서 **상수**에 더 가깝지 않을까 라는 생각에 상수처럼 작명을 했는데, 혹시 로키는 어느 점에서 해당 값들을 상태 값으로 판단하신 건가요?? 궁금합니다!

Rok93 replied:
스프링이 빈으로 관리하는 객체도  인스턴스가 1개로 유지될 뿐이지 일반적인 POJO(Plain Old Java Object)이기 때문에 객체의 상태로 바라보는게 조금 더 일반적인 접근이라고 생각해요. 

빈이 싱글톤 객체이기 때문에 빈 객체가 가지는 값은 상수라고 생각하면 안된다고 생각해요. 👀
그 외에도 상수는 변경가능하면 안되는데, 현재 상수 네이밍을 하신 두 값은 `final` 선언을 하지 않았기 때문에 변경 가능성이 생길 수도 있다고 생각해요.

minSsan replied:
완전 이해했습니다! 제가 너무 스프링 프레임워크에 기반해서 생각했던 것 같아요..~
반영했습니다!

[refactor: 불필요한 상수명 변경](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/a6eb8454ee3edc8aedb2337647693517bb5d0c90)

Rok93 commented on src/main/java/roomescape/global/infrastructure/JwtTokenProvider.java:
```diff
@@ -0,0 +1,38 @@
+
+@Component
+public class JwtTokenProvider {
+    @Value("${security.jwt.token.secret-key}")
```
프로퍼티로 주입받는 방식 👍💯 
추후 application-property에서 환경별로 다른 값을 주입한다거나 혹은 일괄적으로 수정하는 작업등을 쉽게적용할 수 있겠네요.

minSsan replied:
넵! 감사합니다 ~ 🙇🏻‍♀️

Rok93 commented on src/main/java/roomescape/auth/service/MemberAuthService.java:
```diff
@@ -0,0 +1,66 @@
+    public LoginMember getLoginMemberByToken(String token) {
+        String subject = jwtTokenProvider.resolve(token);
+        try {
+            final long userId = Long.parseLong(subject);
```
member를 token으로 변환해주는 역할뿐만 아니라, token을 userId로 변환해주는역할도 jwtTokenProvider에게 위임해보면 어떨까요?

minSsan replied:
[다른 코멘트](https://github.com/woowacourse/spring-roomescape-member/pull/286#discussion_r2086233267)에서 언급했던 것처럼 본래 JwtTokenProvider를 최대한 일반화해서 정의하고 싶어서 역할을 최대한 부여하지 않았었는데요 ..!
오히려 관련 로직이 분산되는 것 같아서 _(관련 내용은 다른 코멘트에서 더 자세히 ...)_ , 코멘트 남겨주신대로 아래와 같이 변환 메서드를 더 추가해줬어요 ~

https://github.com/woowacourse/spring-roomescape-member/blob/7edad65fc9d9dd776ed4776571e5e84352012b40/src/main/java/roomescape/global/infrastructure/JwtTokenProvider.java#L22-L34

https://github.com/woowacourse/spring-roomescape-member/blob/7edad65fc9d9dd776ed4776571e5e84352012b40/src/main/java/roomescape/global/infrastructure/JwtTokenProvider.java#L58-L65

Rok93 commented on src/main/java/roomescape/auth/service/MemberAuthService.java:
```diff
@@ -0,0 +1,66 @@
+                });
+    }
+
+    // TODO: 파라미터 토큰 ? user id?
```
불필요한 주석이라면 제거해주세요. 👀

아마 메서드 파라미터로 어떤 값을 받을 것인지를 고민하신 것이 아닌가 싶은데요. 
저라면 Service단에서 무슨 값인지 알 수 없는 token을 받지 않는 것이 추후 재사용성이 더 높을거라고 생각해서 userId를 받도록할 것 같아요. 🤔

다만 그렇게 변경되게되면 해당 Service는 사실 `MemberService` 와 같이 네이밍해도 무방할 것 같지 않나요? 😃

minSsan replied:
흐음! 🙄 다시 한번 생각해보니 `request / response 쿠키`로 전달되는 **token**은 서비스 계층보다는 더 상위 계층 _(= 컨트롤러와 서블릿 사이 그 어딘가?)_ 에서 관리하는게 더 어울리는 것 같아요
> 특히 **token**이 `HTTP 요청`에서 **클라이언트가 누구인지 구별하기 위해** 사용한다는 목적성으로만 봐도, 해독 과정은 서비스 계층보다는 더 상위 계층에서 처리하는 것이 적합하다고 느껴졌어요

하지만 기존 코드에서는 `JwtTokenProvider` 변경의 여파가 서비스까지 전파 돼요.
> 서비스 내부에서 `JwtTokenProvider`를 통해 토큰을 해석하고 있음

따라서 `JwtTokenProvider`로 유저의 id _(JWT token subject)_ 를 추출하는 작업을 Interceptor / Resolver와 같은 상위 계층에서 처리하고, 관련된 변경 여파가 서비스에 전파되지 않도록 _(= 로키가 말한 높은 재사용성의 연장선 느낌?!)_ 하는 것이 좋을 것 같다는 생각이 들었어요 ~

---

[fix: 어드민 reservation 생성 요청시 회원 존재 검증(with DB) 미포함](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/809a2fca4d38693d623b3a779e37150e85aa7d19)

Rok93 commented on src/main/java/roomescape/auth/service/MemberAuthService.java:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    public List<MemberIdAndNameResponse> getAllMemberNames() {
```
DTO 객체 네이밍짓는게 항상 애매한 경우가 많은데요. 지금의 네이밍은 가능한 지양하시면 좋을 것 같아요. 😃
예를들어 추후 제공해야할 응답값이 더 늘어나게된다면 어떻게될까요...?

minSsan replied:
로키 말씀에 동의합니다..! 제가 한거긴 하지만 이름 지으면서도 참 구리다..고 생각했습니다 ㅜ 💦
조금 더 일반화해서 `MemberBasicInfoResponse`로 이름을 변경했어요 ..!

[refactor: DTO 네이밍 변경 - MemberIdAndNameResponse -> MemberBasicInfoResp…](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/4dca0cff9aac21b9b2385ce1e18da2e6881cdc96)

Rok93 commented on src/main/java/roomescape/auth/service/dto/request/AdminSignupRequest.java:
```diff
@@ -0,0 +1,19 @@
+) {
+    public Member toEntity() {
+        return new Member(0L, name, Role.ADMIN.name(), email, password);
+    }
```
토끼의 의견에 동의합니다.  
(사유가 같은건 아니라서 제 생각을 남겨보자면...!)
non-nullable type을 지원하는 언어라면 `0L` 하드코딩도 나쁘지 않은 선택이될 수도 있지만, 이런 타입 시스템을 지원하지 않는 java에서 굳이 0으로 하드코딩을 해줄 필요는 없다고 생각해요.

minSsan replied:
앞서 다른 크루를 통해서도 해당 코드에 대해서 피드백을 받았는데요 🤔
> 코드를 처음 읽어보는 사람 입장에서 `DB에 0이라는 id 값으로 저장될 수 있다`고 오해할 여지가 있다고 지적했거든요.

저도 사실 처음에는 0이 아닌 `null`로 정의를 했었다가, **id의 타입을 `Long` 타입으로 통일하기 위해 0으로 초기화해도 괜찮겠다**고 생각했어요.
- id 생성의 책임은 어차피 DB에게 있고, 1부터 시작해서 생성되기 때문에 0으로 초기화 하더라도 오해의 여지가 없을거라고 생각했거든요.

하지만 역으로 생각해서, **id 필드의 존재 이유** 측면에서 영속화되지 않은 엔티티**에 대해서는 id 값을 참조할 일도 없을 것** 같아요. 그렇기 때문에 `id 필드의 접근 안정성을 위해 Long 타입으로 통일한다`는 주장이 큰 설득력이 없을 수 있겠다는 생각이 들었어요 🤪

내용 반영했습니다!

[refactor: 영속화 되지 않은 Entity 초기 id 값을 null로 변경](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/fd2a1cc7a39894c0ee66ed58eb59089c4f92cb0a)

Rok93 commented on src/main/java/roomescape/global/infrastructure/TokenCookieProvider.java:
```diff
@@ -0,0 +1,22 @@
+            }
+        }
+        throw new MemberUnauthorizedException();
+    }
```
cookie가 없거나  cookie에 token 값이 없는경우 MemberUnauthorizedException 예외를 던지고있는데요. 
예외 메시지가 `등록되어 있지 않은 유저 정보입니다.` 인것으로 보여요. 

오류 메시지를 처음 본 입장에서는 조금 혼동의 여지가 있을 것 같은데요. 
조금 더 적절한 예외 메시지를 던져주면 어떨까요? 😃

<img width="535" alt="image" src="https://github.com/user-attachments/assets/1f9b6072-adf7-40c8-a8df-28ebad8b976d" />

minSsan replied:
쿠키가 존재하지 않아서 예외가 발생하는 상황은 **로그인이 필요한 서비스에 로그인 없이 접근한 경우**일 것 같아요!
`"로그인이 필요한 서비스"` 임을 명시하도록 메시지를 변경했어요 ~

[feat: 401 에러 문구 변경](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/c8c449a97d1497632029e15518d1ca908b7d576e)

Rok93 commented on src/main/resources/schema.sql:
```diff
@@ -16,12 +26,13 @@ CREATE TABLE theme
+    FOREIGN KEY (member_id) REFERENCES member (id),
     FOREIGN KEY (time_id) REFERENCES reservation_time (id),
     FOREIGN KEY (theme_id) REFERENCES theme (id)
 );
```
<img width="1140" alt="image" src="https://github.com/user-attachments/assets/28e70a21-f38f-46ac-ba05-b1244f208945" />

실제 애플리케이션을 실행하여 테스트를 해보고싶은데, 더미데이터가 하나도 없어서 테스트를해보기가 조금 어려운 것 같아요. 🥲
물론 실제로는 dev, stg와 같은 테스트 환경들이 존재하고 여기에 더미 데이터를 미리 넣어두거나 혹은 (유저 플로우대로) 직접 생성하는 방식을 통해서 할 수 도 있지만, 저희는 현재 별도 개발 환경이 없기 때문에 아쉽지만 해당 sql 파일에 더미 데이터를 추가해주셨으면 합니다. 🙏

minSsan replied:
아앗 죄송합니다 .. 🥺 사실 더미 데이터를 넣었었는데 `data.sql`로 더미 데이터 넣는 순간 테스트가 다 터져버려서 🤯.. 
제출 전까지 해결 방법을 찾아보려고 했는데, 잘 모르겠어서 결국 더미 데이터 없이 제출했었어요 ㅜ

일단 `data.sql` 추가해서 커밋 해두었습니다..!!!

[setup(data.sql): 더미 데이터 추가 (테스트 깨짐)](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/e677960e2f88221a2754f9a482c0ee7125cab0e6)

Rok93 commented on src/main/java/roomescape/theme/repository/JdbcThemeRepository.java:
```diff
@@ -80,6 +76,7 @@ public Optional<Theme> findById(Long id) {
         }
     }
 
+    // TODO: (모든 Repository, Service에 적용) try-catch 삭제 및 Service에서 DataAccessException 핸들링하도록
```
TODO 주석들이 많은 것 같은데, 아직 진행이 되지 않을걸까요? 
혹시 진행한 작업이라면 주석 제거 부탁드릴게요. 🙏

minSsan replied:
관련해서 더 고민해보고 적용 여부를 결정하려고 했었는데, 시간이 부족해서 고민을 더 해보지 못했어요 💦

`@Repository` 어노테이션이 데이터 접근 관련 예외를 **어느 데이터 접근 구현체를 사용하더라도 일관성있게 처리해주는 것**으로 알고 있는데요, 현재 코드에서는 **Repository 구현체 내부에서 예외를 다 처리**해주고 있어요.
그렇다면 `@Repository` 어노테이션의 장점을 활용하지 못하고 있는게 아닌가 싶은 생각이 들었거든요.
> 만약 서비스에서 레파지토리 예외를 핸들링하면, `@Repository` 구현체가 변경되더라도 예외 핸들링 코드를 수정할 필요가 없음

그래서 레파지토리 구현체 내부에 있는 예외 핸들링 코드를 외부(= 서비스 계층)로 빼는 걸 고민해봤었는데요, 관련해서 로키의 의견도 한번 들어보고 싶어요..!

Rok93 replied:
사실 일관성있게  예외를 처리할 수 있는 것은 지금도 적용할 수 있지만... 

 `@Repository` 애너테이션은 빈을 등록하기위한 용도 (+ 계층을 명시적으로  구분)로 밖에 사용하지 못하는 점은 조금 아쉽지만, 저는 Repository 계층에서 어떤 단일 값이 존재하는지 아닌지에 대해서 Optional로 제공해주고 Service단에서는 존재 여부에 따라 적절히 취사선택 할 수있도록 제공해주는 것이 좋다고 생각해요. 

그렇지 않았을 경우 service 단에서 단일값 조회 로직은 매번 try-catch 문을 감싸서 처리를 해줘야하기 때문에 Repository에서 Optional로 결과를 감싸서 제공하는 방식에 비해 불필요한 로직들이 반복적으로 많이 생길 것 같아요. 👀 이런 부분의 트레이드오프를 생각했을 때, 적어도 단일 값 조회 정도는 Repository 단에서 예외핸들링을해서 결과를 Optional로 제공해주는 정도의 처리는 해줘도 좋지 않나 생각해요. 😃

minSsan replied:
그러고보니 **여러 개의 서비스 레이어 객체에서 에러를 핸들링**하면, _중복되는 예외 처리도 분산돼서 관리될 것_ 이고 그로 인해 _예외 요구사항이 바뀔 때마다 작업량이 오히려 늘어날 것_ 같네용 ..!

로키 말씀대로 `단일 조회 예외`는 **레파지토리 내부에서 일괄적으로 처리**하는게 좋을 것 같아요~
TODO 주석은 모두 삭제했어요

Rok93 commented on src/main/java/roomescape/reservation/entity/Reservation.java:
```diff
@@ -9,38 +9,25 @@
-        validateFields(0L, name, date, time, themeId);
-        LocalDateTime now = LocalDateTime.now();
+    public boolean isBefore(LocalDateTime compare) {
         LocalDateTime dateTime = LocalDateTime.of(date, time.getStartAt());
```
`LocalDateTime.now()`와 같이 제어하기 어려운 코드에 의존하게되면 추후 테스트코드 작성이 어려워질 것 같은데요. 🤔

레벨1에서 학습했던 것처럼 테스트하기 어려운 요소를 최대한 밖으로 꺼내서 테스트 영역을 확보해보면 어떨까요?

minSsan replied:
어이쿠야.. 지난 리뷰요청 때 해주신 질문인데 PR이랑 같이 답변 남긴다는걸 깜빡했네요 😵‍💫

---

관련해서 지난 미션에서도 고민했었는데요..!
https://github.com/woowacourse/spring-roomescape-admin/pull/319#discussion_r2058177595

### 코멘트 요약
> - 외부 의존성을 갖는 서비스 _(ex DB, 메시징 서비스 ...)_ 가 아닌, 단순 날짜 데이터를 테스트 편리성을 위해 분리할 필요성을 크게 느끼지 못했다
> - 테스트에서 날짜를 제어하기 위해서 프로덕션 코드(?)에 `DateTimeGenerator` 의존성이 추가되는 것이 오히려 복잡도가 늘어나는 것처럼 보였다
>     - Fake 테스트를 하려면 `DateTimeGenerator`를 인터페이스로 추상화 해야하는데, 코드 복잡도가 오히려 늘어나는 느낌이 들었다
> - 만약 추상화를 적용하지 않는다면, Stub 테스트로 리턴 값을 현재 날짜(`LocalDateTime.now()`)로 고정할 수 있을 것 같다. 그렇다면 [현재 테스트 코드](https://github.com/woowacourse/spring-roomescape-member/pull/286/files#diff-a0db0a41a592c9a6f2f76f6849aa293483217359d6a4431bdf992cf2e507584aR75-R89)와 무슨 차이가 있는거지??

DB같은 **외부 서비스에 의존하는 객체**에 대해서는 `의존성 주입`을 통해 의존성을 표시하는 이점이 크다고 생각하지만, 지금처럼 **날짜같은 단순 데이터**에 대한 의존성을 `의존성 주입`으로 표시하는 이점이 크게 안 와닿았던 것 같아요 🫨

그리고 스스로 납득되지 않는 부분은 코드에 반영하지 않는게 맞다고 생각해서 변경하지 않았어요.. 🥹

혹시 제가 미처 고려하지 못한 부분이나 간과하고 있는 부분이 있을까요?? 로키는 이에 대해서 어떻게 생각하시나요??

minSsan committed:
refactor: JwtTokenProvider 의 Claims 결정 책임을 외부로 변경

Rok93 commented on src/main/java/roomescape/auth/service/MemberAuthService.java:
```diff
@@ -0,0 +1,75 @@
+    public LoginResponse login(LoginRequest request) {
+        Member member = memberRepository.findByEmailAndPassword(request.email(), request.password())
+                .orElseThrow(MemberUnauthorizedException::new);
+        Claims claims = Jwts.claims()
+                .subject(member.getId().toString())
+                .add("role", member.getRole().name())
+                .build();
```
개인적으로 생각하기에는 `io.jsonwebtoken`와 관련된 의존성은 모두 JWT를 생성하기위한 의존성이라 생각하는데요. 
가능한 이런 의존성들은 JwtTokenProvider에게 위임해보면 어떨까요? 

```java
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
```

minSsan replied:
음.. 사실 `JwtTokenProvider`를 최대한 일반화해서 사용하고 싶어서 이렇게 정의했었어요~
> `Claims`, `Subject` 타입의 결정권을 외부로 넘김

이렇게 정의하면 현재 Auth에 필요한 JWT 토큰 생성 뿐만 아니라, 더 다양한 상황에서 토큰을 발급받고 활용할 수 있지 않을까 싶어서요..!

하지만 이렇게 정의하다보니 **Claim을 생성 / 해독하는 코드가 분산되는 문제**가 계속 발생했어요. _(= `Interceptor`, `Resolver`, `Service`에 관련 로직이 분산되어서 존재함)_

그래서 기존에 세웠던 '최대한 일반화하기' 라는 목표에 대해서 다시 생각해봤어요. 
그리고 `JwtTokenProvider`를 일반화해서 정의하기보다는, 만약 다른 요구사항에도 토큰 생성이 필요하게 된다면 `TokenProvider`를 추상화해서 여러 구현체를 가져도 되겠다는 생각이 들었어요.
- 이때 어떤 객체에 대해서 claim을 결정할 것인지 / subject 타입은 무엇인지도 제네릭을 통해 충분히 추상화가 가능하지 않을까 하는 생각이 들었습니다  

이처럼 기존 코드의 방식 말고도 `JwtTokenProvider`를 다양한 상황에서 적용할 수 있는 방안이 충분히 존재하는 것 같아서, Claim 생성 및 해독하는 책임을 모두 `JwtTokenProvider` 내부로 옮겼어요 ~

[fix: 어드민 reservation 생성 요청시 회원 존재 검증(with DB) 미포함](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/809a2fca4d38693d623b3a779e37150e85aa7d19)

Rok93 replied:
> 하지만 이렇게 정의하다보니 Claim을 생성 / 해독하는 코드가 분산되는 문제가 계속 발생했어요. (= Interceptor, Resolver, Service에 관련 로직이 분산되어서 존재함)

직접 불편함을 느끼고 그 불편함을 해결하기위한 방향으로 코드 수정 하는 과정이 잘 보여서 너무 좋네요. 😊👍💯

Rok93 commented on src/main/java/roomescape/global/config/WebMvcConfiguration.java:
```diff
@@ -0,0 +1,36 @@
+package roomescape.global.config;
```
global인 경우와 아닌경우의 차이는 뭔가요? 👀
개인적으로 global이라는 패키지가 조금 애매한 성격을 띄고있다고 생각해요. 
현재 global 패키지 하위에 config, exception, infrastructure 등의 패키지가위치하는데요. 

일반적으로 설정(config)쪽은 대다수가 전역적으로 적용되는 경우가 많다보니 config 자체가 global에 적용된다는 성격을 가지고 있을 것 같아서 그냥 config라고 뒀어도 충분히 비슷하게 이해하지 않을까 싶어요.

minSsan replied:
로키가 말씀해주신 것처럼 `global` 패키지는 **클래스가 전역으로 사용되는지**를 구분하려는 목적이었어요
사실 이번 미션에서 패키지 분류 기준을 어떻게 둬야할지 모호하게 느껴져서, 패키지 분리가 특히 조금 어렵게 느껴졌어요 💦

현업 개발 경험이 있는 입장에서 봤을 때, `config`라는 네이밍 만으로도 `global` 패키지의 의미가 내포되어 있다면 로키 말씀대로 `config`는 바깥으로 빼도 될 것 같아요 !

의견 남겨주셔서 감사합니다 🙇🏻‍♀️

Rok93 commented on src/main/java/roomescape/global/infrastructure/JwtTokenProvider.java:
```diff
@@ -0,0 +1,36 @@
+package roomescape.global.infrastructure;
```
추가적으로 token과 관련된 로직은 `infrastructure` 라는 패키지 하위에있는데요. 
밍곰이 생각하시는 `infrastructure`는 어떤 요소들이 포함되어야한다고 생각하시나요? 😃 

저는 개인적으로 infrastructure 패키지 하위의 객체들이  `authentication`을 위한 용도라고 생각해서 이와 비슷한 네이밍의 패키지 하위에 둘 것 같아요. 😃

Rok93 replied:
이 부분도 생각 남겨주시면 좋을 것 같아요. 😃

minSsan replied:
`infrastructure` 영역은 **외부 시스템과 상호작용하는 구현체**들의 영역이라고 생각해요
즉, **도메인 로직이 외부 시스템에 의존하지 않도록** 중개하는 느낌? 정도로 이해했습니다 🤔

#### `JwtTokenProvider` & `AuthTokenCookieProvider`를 infrastructure에 포함시킨 이유
`JwtTokenProvider`나 `AuthTokenCookieProvider`도 각각 JWT / 쿠키 _(= 저는 이 두 가지는 도메인 영역에는 해당하지 않고, 외부 요인으로 볼 수 있다고 생각했습니다)_ 와 상호작용하는 작업이기 때문에 `infrastructure`에 뒀어요.

그리고 인증/인가에서 사용되는 두 객체는 특정 도메인에 종속되어있지 않다고 생각했습니다..! 
> 인증/인가가 필요한 요청을 처리하는 어느 곳에서든 참조될 수 있으니까???

그래서 `auth` 패키지에 두지 않고 `global`이라는 패키지에 위치시켰어욥

Rok93 replied:
지금의 구조에서는 조금 안맞을 수도 있지만 저는 `JwtTokenProvider`라거나 `AuthTokenCookieProvider` 같은 경우에는 결국 직접적으로 service단 같은 곳에서 안쓰고 인터셉터나 아규먼트 리졸버 정도에서 사용했을 것 같아요. 

그랬다면 이 두 클래스 파일은 config 하위에 위치시켰을 것 같고, 이 둘은 인증과 관련된 개념들이니 auth 
즉 config.auth 패키지 하위에 위치시켰을 것 같아요. 😃

생각해보면 config(= 설정)는 전역적으로 적용되는 경우가 많다보니 별도로 global 이라는 네이밍이 필요한가 싶기도하구요. 🤔

사실 이런 패키지 네이밍 관련해서는 팀적인 합의가 중요한거기 때문에 저의 의견도 절대로 정답이 아니라고생각해요. 😉

Rok93 reviewed (changes requested):
밍곰 4~6단계도 잘 수행해주셨네요. 👍👍 
몇몇 코멘트 남겨두었는데, 확인 부탁드릴게요. 🙏

minSsan committed:
chore: TokenCookieProvider 네이밍 변경

minSsan committed:
chore: LoginController -> AuthController 네이밍 변경

minSsan committed:
feat(AuthController): 로그아웃 기능 추가

minSsan committed:
feat: Reservation 생성 요청시 memberId 검증 생략 - dto에 LoginMember 포함

minSsan committed:
chore: LoginMember 패키지 변경

minSsan committed:
fix: 어드민 reservation 생성 요청시 회원 존재 검증(with DB) 미포함

- Claims 생성 책임을 JwtTokenProvider 내부로 이동
- service에서 LoginMember 생성시 회원 존재 검증 포함 (with DB)
- JwtTokenProvider: Claim getter 추가

minSsan committed:
refactor(JwtTokenProvider): 불필요한 메서드 삭제

minSsan committed:
refactor: MemberAuthService -> MemberService 네이밍 변경

minSsan committed:
feat: mockito 의존성 삭제 및 JwtTokenProvider Stub 상속으로 구현

minSsan committed:
refactor: 영속화 되지 않은 Entity 초기 id 값을 null로 변경

minSsan committed:
fix(FakeThemeRepository): id 미지정시 테스트 깨짐 - id 없을시 자동 생성 로직 추가

minSsan committed:
test: Fake 레파지토리 객체 save id 자동 생성 로직 추가

minSsan committed:
refactor: DTO 네이밍 변경 - MemberIdAndNameResponse -> MemberBasicInfoResponse

minSsan committed:
feat: 401 에러 문구 변경

minSsan committed:
feat: 로그인 시 존재하지 않는 회원정보 오류 status 변경 - 401 -> 404

minSsan committed:
setup(data.sql): 더미 데이터 추가 (테스트 깨짐)

Rok93 reviewed (changes requested):
밍곰 안녕하세요! 
현재 깨지는 테스트가 꽤 있어서 해당 테스트들 통과하도록 수정해서 다시 리뷰 요청 부탁드릴게요. 🙏

여러번 테스트 돌려봤는데도 테스트가 깨지는 것으로 보이는데 그럼에도 제 로컬에서만 발생하는 현상일 수 있으니 혹시 밍곰의 환경에서 잘 돌아간다면 저한테 말씀해주세요. 

<img width="1604" alt="image" src="https://github.com/user-attachments/assets/79d783b6-509f-44f5-aba5-3a9482575541" />

minSsan committed:
setup(test/application.properties): 테스트 환경 설정 - data.sql 비활성화

minSsan committed:
fix: data sql auto increment 적용 - id 지정 삭제

Rok93 commented on src/main/java/roomescape/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,43 @@
+    public void login(@RequestBody @Valid LoginRequest request, HttpServletResponse response) {
+        LoginResponse loginResponse = service.login(request);
+        Cookie cookie = authTokenCookieProvider.generate(loginResponse.token());
+        response.addCookie(cookie);
```
ㅇㅇㅇ

minSsan replied:
ㅇ_ㅇ ?!?!

Rok93 replied:
ㅋㅋㅋㅋㅋ 죄송합니다... 😅

Rok93 commented on src/main/java/roomescape/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,43 @@
+        LoginResponse loginResponse = service.login(request);
+        Cookie cookie = authTokenCookieProvider.generate(loginResponse.token());
+        response.addCookie(cookie);
+    }
```
`HttpServletResponse`를 사용하는 방식 대신에 아래와 같이 ResponseEntity의 header를 통해 cookie를 지정하는 방법은 어떨까요? 

```
    return ResponseEntity.ok()
            .header(HttpHeaders.SET_COOKIE, "token", "tokenValue")
            .build();
```

minSsan replied:
기존에 사용했던 `Cookie`로 코멘트 주신 내용 반영하려다보니 제약사항이 조금 많아서 _(`toString` 지원 안 해줌)_ , `ResponseCookie`로 변경해서 반영했어요~

[feat: Cookie -> ResponseCookie 변경](https://github.com/woowacourse/spring-roomescape-member/pull/286/commits/ec95bcd08b0c090bd5325d8f88ab217f89a3bf9b)

Rok93 commented on src/main/java/roomescape/auth/repository/JdbcMemberRepository.java:
```diff
@@ -0,0 +1,99 @@
+                    resultSet.getString("role"),
+                    resultSet.getString("email"),
+                    resultSet.getString("password"));
+    private final NamedParameterJdbcTemplate jdbcTemplate;
```
사소할 수 있지만 상수와 상태는 줄바꿈을 통해서 구분하기 쉽게해주면 코드 파악이 조금 더 쉬울 것 같아요. 😃 

```suggestion
                    resultSet.getString("password"));
                    
    private final NamedParameterJdbcTemplate jdbcTemplate;
```

Rok93 reviewed (changes requested):
밍곰 피드백 반영 잘해주셨어요. 👍👍 
지금 당장 머지해도 될 정도로 전체적으로 잘 구현해주셨지만 혹시나 수정하신 부분이있을까 싶어 마지막으로 Request Changed 드려요. 😃

다음 리뷰 요청 때는 방탈출 사용자 예약 미션은 이만 종료하도록하겠습니다. 🤗
마지막으로 코멘트들 확인 및 수정할 부분이있으시다면 수정해주신 뒤 리뷰요청주세요. 
추가적으로 질문 남겨주신 부분들에대한 제 생각도 남겨두었습니다. 😉

minSsan committed:
refactor(LoginTest): 중복되는 로그인 로직 분리

minSsan committed:
feat: 아이디, 비밀번호 검증 쿼리 변경 - email만 비교

minSsan committed:
feat(schema.sql): reservation NOT NULL 제약 조건 추가 + member email UNIQUE

minSsan committed:
refactor: 불필요한 상수명 변경

minSsan committed:
chore: 불필요한 TODO 주석 삭제

minSsan committed:
feat: Cookie -> ResponseCookie 변경

minSsan committed:
fix: 데이터 스키마 변경으로 인해 테스트 깨짐 - not null 필드 삽입

minSsan committed:
chore: 공백 추가

Co-authored-by: Gyeonglok Kim <goodboy302@naver.com>

minSsan committed:
chore(JwtTokenProvider): getClaims 메서드 private으로 변경

minSsan committed:
refactor: config 패키지 위치 변경

minSsan reviewed (commented):
안녕하세요 로키, 남겨주신 코멘트 내용 반영해서 다시 리뷰요청 보내드렸어요~
마지막까지 잘 부탁드립니다..! 🙇🏻‍♀️

그리고 이전 리뷰에서 남겨주신 질문에 대한 답변입니다!
https://github.com/woowacourse/spring-roomescape-member/pull/286#discussion_r2092879258

Rok93 reviewed (approved):
밍곰 마지막까지 피드백 반영 잘해주셨어요. 👍👍 
방탈출 사용자 예약 미션은 이만 종료하도록할게요. 🎉🎉 
고생 많으셨고 남은 미션도 화이팅입니다. 💪💪💪
